lightning/src/routing/router.rs

```diff
@@ -1535,17 +1550,93 @@ where L::Target: Logger {
+	let network_channels = network_graph.channels();
+	let network_nodes = network_graph.nodes();
+
+	for path in route.paths.iter_mut() {
+		let mut shadow_ctlv_expiry_delta_offset: u32 = 0;
+
+		// Choose the last publicly known node as the starting point for the random walk
+		if let Some(starting_hop) = path.iter().rev().find(|h| network_nodes.contains_key(&NodeId::from_pubkey(&h.pubkey))) {
+			let mut cur_node_id = NodeId::from_pubkey(&starting_hop.pubkey);
+
+			// Init PRNG with path nonce
+			let mut path_nonce = [0u8; 12];
+			path_nonce.copy_from_slice(&cur_node_id.as_slice()[..12]);
+			let mut prng = ChaCha20::new(random_seed_bytes, &path_nonce);
+			let mut random_path_bytes = [0u8; ::core::mem::size_of::<usize>()];
+
+			// Pick a random path length in [1 .. 3]
+			prng.process_in_place(&mut random_path_bytes);
+			let random_walk_length = usize::from_be_bytes(random_path_bytes).wrapping_rem(3).wrapping_add(1);
+
+			for _random_hop in 0..random_walk_length {
+				if let Some(cur_node) = network_nodes.get(&cur_node_id) {
+					// Randomly choose the next hop
+					prng.process_in_place(&mut random_path_bytes);
+					if let Some(random_channel) = usize::from_be_bytes(random_path_bytes).checked_rem(cur_node.channels.len())
+						.and_then(|index| cur_node.channels.get(index))
+						.and_then(|id| network_channels.get(id)) {
+							random_channel.as_directed_from(&cur_node_id).map(|(dir_info, next_id)| {
+								dir_info.direction().map(|channel_update_info|
+									shadow_ctlv_expiry_delta_offset = shadow_ctlv_expiry_delta_offset
+										.checked_add(channel_update_info.cltv_expiry_delta.into())
+										.unwrap_or(shadow_ctlv_expiry_delta_offset));
+								cur_node_id = *next_id;
+							});
+						}
+				}
+			}
+		} else {
+			// If the entire path is private, choose a random offset from multiples of
+			// MEDIAN_HOP_CLTV_EXPIRY_DELTA
+			let mut prng = ChaCha20::new(random_seed_bytes, &[0u8; 8]);
+			let mut random_bytes = [0u8; 4];
+			prng.process_in_place(&mut random_bytes);
+			let random_walk_length = u32::from_be_bytes(random_bytes).wrapping_rem(3).wrapping_add(1);
+			shadow_ctlv_expiry_delta_offset = random_walk_length.wrapping_mul(MEDIAN_HOP_CLTV_EXPIRY_DELTA);
+		}
+
+		// Limit the total offset to reduce the worst-case locked liquidity timevalue
+		const MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET: u32 = 3*144;
+		shadow_ctlv_expiry_delta_offset = cmp::min(shadow_ctlv_expiry_delta_offset, MAX_SHADOW_CLTV_EXPIRY_DELTA_OFFSET);
+
+		// Limit the offset so we never exceed the max_total_cltv_expiry_delta. To improve plausibility,
+		// we choose the limit to be the largest possible multiple of MEDIAN_HOP_CLTV_EXPIRY_DELTA.
+		let mut max_path_offset = payment_params.max_total_cltv_expiry_delta
+			.checked_sub(path.iter().map(|h| h.cltv_expiry_delta).sum())
+			.unwrap_or(0);
+		max_path_offset = max_path_offset
+			.saturating_sub(max_path_offset.wrapping_rem(MEDIAN_HOP_CLTV_EXPIRY_DELTA))
```

Comment:
A lot of the code here and above would be substantially easier to read without the functional-style math. eg the docs for `wrapping_rem` even explicitly states that its identical to `%`, and above there's a good chunk of math that trivially cannot overflow.

Reply:
Hm, sorry for that, I got into the habit of using the 'functional style' consistently throughout, especially when some closely connected parts of the code require it. I should reconsider that since I agree that it's sometimes not as easy to read. I now further simplified the parts in question with https://github.com/lightningdevkit/rust-lightning/commit/e2efb0c4f01db56b4fd5d0b8898d820b55747ecf.